<a href="https://colab.research.google.com/github/karywnl/kaggle/blob/main/data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd

In [3]:
import os
from google.colab import drive

drive.mount("/content/drive")
os.environ["KAGGLE_CONFIG_DIR"] = "/content/drive/MyDrive/Kaggle"

!chmod 600 "/content/drive/MyDrive/Kaggle/kaggle.json"

Mounted at /content/drive


In [4]:
!kaggle datasets download -d "maxhorowitz/nflplaybyplay2009to2016"
!unzip nflplaybyplay2009to2016.zip

Dataset URL: https://www.kaggle.com/datasets/maxhorowitz/nflplaybyplay2009to2016
License(s): unknown
 99% 271M/274M [00:00<00:00, 423MB/s]
100% 274M/274M [00:00<00:00, 486MB/s]
Archive:  nflplaybyplay2009to2016.zip
  inflating: NFL Play by Play 2009-2016 (v3).csv  
  inflating: NFL Play by Play 2009-2017 (v4).csv  
  inflating: NFL Play by Play 2009-2018 (v5).csv  


In [5]:
nfl_data = pd.read_csv("NFL Play by Play 2009-2017 (v4).csv")

/tmp/ipykernel_318/2226537327.py:1: DtypeWarning: Columns (25,51) have mixed types. Specify dtype option on import or set low_memory=False.
  nfl_data = pd.read_csv("NFL Play by Play 2009-2017 (v4).csv")


So, there is a concept of low memory in pandas, when the dataset is huge. The pandas doesn't read them all at once. Instead it will read them in chunk. So, the problem with this approach is different chunks can be seen as a different datatype. That's why we get the mixed types warning in the above cell. If low_memory = False, then whole dataset is read at once and we may get no warning. Usually if the size of the dataset is like > 50 MB.

In [6]:
nfl_data = pd.read_csv("NFL Play by Play 2009-2017 (v4).csv", low_memory=False)
nfl_data.head()

,Date,GameID,Drive,qtr,down,time,TimeUnder,TimeSecs,PlayTimeDiff,SideofField,...,yacEPA,Home_WP_pre,Away_WP_pre,Home_WP_post,Away_WP_post,Win_Prob,WPA,airWPA,yacWPA,Season
0,2009-09-10,2009091000,1,1,NaN,15:00,15,3600.0,0.0,TEN,...,NaN,0.485675,0.514325,0.546433,0.453567,0.485675,0.060758,NaN,NaN,2009
1,2009-09-10,2009091000,1,1,1.0,14:53,15,3593.0,7.0,PIT,...,1.146076,0.546433,0.453567,0.551088,0.448912,0.546433,0.004655,-0.032244,0.036899,2009
2,2009-09-10,2009091000,1,1,2.0,14:16,15,3556.0,37.0,PIT,...,NaN,0.551088,0.448912,0.510793,0.489207,0.551088,-0.040295,NaN,NaN,2009
3,2009-09-10,2009091000,1,1,3.0,13:35,14,3515.0,41.0,PIT,...,-5.031425,0.510793,0.489207,0.461217,0.538783,0.510793,-0.049576,0.106663,-0.156239,2009
4,2009-09-10,2009091000,1,1,4.0,13:27,14,3507.0,8.0,PIT,...,NaN,0.461217,0.538783,0.558929,0.441071,0.461217,0.097712,NaN,NaN,2009


In [7]:
nfl_data.isnull().sum()

,0
Date,0
GameID,0
Drive,0
qtr,0
down,61154
...,...
Win_Prob,25009
WPA,5541
airWPA,248501
yacWPA,248762


.sum() here is by default goes by column-wise because its axis=0 by default. If axis=1, then it goes row-wise.

*   axis = 0 : Null count by row - col b col and col on the index
*   axis = 1 : Null count by col - row by row and row on the index

In [8]:
nfl_data.isnull()

,Date,GameID,Drive,qtr,down,time,TimeUnder,TimeSecs,PlayTimeDiff,SideofField,...,yacEPA,Home_WP_pre,Away_WP_pre,Home_WP_post,Away_WP_post,Win_Prob,WPA,airWPA,yacWPA,Season
0,False,False,False,False,True,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
407683,False,False,False,False,True,False,False,False,False,False,...,True,True,True,True,True,True,False,True,True,False
407684,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
407685,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
407686,False,False,False,False,False,False,False,False,False,False,...,True,False,False,False,False,False,False,True,True,False


nfl_data.isnull() creates a dataframe, with booleans in each cell, Indicating that whethere that cell is True or False.

nfl_data.isnull().sum() goes column-wise and tells how many null values each column has and more importantly it returns a Series.

Series have same property just like list and it can also be sliced. For eg : we could get select particulars and see how many null are there in those range of rows.

The below code demonstrates, how to select the null count from first 10 rows column-wise.

In [9]:
nfl_data.isnull().sum()[0:10]

,0
Date,0
GameID,0
Drive,0
qtr,0
down,61154
time,224
TimeUnder,0
TimeSecs,224
PlayTimeDiff,444
SideofField,528


In [10]:
nfl_data.isnull().sum()

,0
Date,0
GameID,0
Drive,0
qtr,0
down,61154
...,...
Win_Prob,25009
WPA,5541
airWPA,248501
yacWPA,248762


***Imputation : You make an educated guess of what the missing value would be and fill them, instead of just dropping them.***

In [18]:
nfl_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407688 entries, 0 to 407687
Columns: 102 entries, Date to Season
dtypes: float64(33), int64(31), object(38)
memory usage: 317.3+ MB


colab usually truncates the output if it is too long like in this case there are 102 columns. we need to use nfl_data.info(verbose=True) to output the whole output.

In [22]:
!kaggle datasets download -d "aparnashastry/building-permit-applications-data"
!unzip building-permit-applications-data.zip

Dataset URL: https://www.kaggle.com/datasets/aparnashastry/building-permit-applications-data
License(s): DbCL-1.0
building-permit-applications-data.zip: Skipping, found more recently modified local copy (use --force to force download)
Archive:  building-permit-applications-data.zip
  inflating: Building_Permits.csv    
  inflating: DataDictionaryBuildingPermit.xlsx  


In [24]:
sf_permits = pd.read_csv("Building_Permits.csv")
sf_permits.head()

/tmp/ipykernel_318/192664614.py:1: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  sf_permits = pd.read_csv("Building_Permits.csv")


,Permit Number,Permit Type,Permit Type Definition,Permit Creation Date,Block,Lot,Street Number,Street Number Suffix,Street Name,Street Suffix,...,Existing Construction Type,Existing Construction Type Description,Proposed Construction Type,Proposed Construction Type Description,Site Permit,Supervisor District,Neighborhoods - Analysis Boundaries,Zipcode,Location,Record ID
0,201505065519,4,sign - erect,05/06/2015,0326,023,140,NaN,Ellis,St,...,3.0,constr type 3,NaN,NaN,NaN,3.0,Tenderloin,94102.0,"(37.785719256680785, -122.40852313194863)",1380611233945
1,201604195146,4,sign - erect,04/19/2016,0306,007,440,NaN,Geary,St,...,3.0,constr type 3,NaN,NaN,NaN,3.0,Tenderloin,94102.0,"(37.78733980600732, -122.41063199757738)",1420164406718
2,201605278609,3,additions alterations or repairs,05/27/2016,0595,203,1647,NaN,Pacific,Av,...,1.0,constr type 1,1.0,constr type 1,NaN,3.0,Russian Hill,94109.0,"(37.7946573324287, -122.42232562979227)",1424856504716
3,201611072166,8,otc alterations permit,11/07/2016,0156,011,1230,NaN,Pacific,Av,...,5.0,wood frame (5),5.0,wood frame (5),NaN,3.0,Nob Hill,94109.0,"(37.79595867909168, -122.41557405519474)",1443574295566
4,201611283529,6,demolitions,11/28/2016,0342,001,950,NaN,Market,St,...,3.0,constr type 3,NaN,NaN,NaN,6.0,Tenderloin,94102.0,"(37.78315261897309, -122.40950883997789)",144548169992


What percentage of the values in the dataset are missing? Your answer should be a number between 0 and 100. (If 1/4 of the values in the dataset are missing, the answer is 25.)

In [38]:
(sf_permits.isnull().sum().sum() / np.prod(sf_permits.shape)) * 100

np.float64(26.26002315058403)

If you removed all of the rows of sf_permits with missing values, how many rows are left?

In [42]:
sf_permits.dropna(axis=0).shape[0]

0

Now try removing all the columns with empty values.

*   Create a new DataFrame called sf_permits_with_na_dropped that has all of the columns with empty values removed.
*   How many columns were removed from the original sf_permits DataFrame? Use this number to set the value of the dropped_columns variable below.



In [43]:
sf_permits_na_dropped = sf_permits.dropna(axis=1)
dropped_columns = sf_permits.shape[1] - sf_permits_na_dropped.shape[1]

Try replacing all the NaN's in the sf_permits data with the one that comes directly after it and then replacing any remaining NaN's with 0. Set the result to a new DataFrame sf_permits_with_na_imputed.

In [44]:
sf_permits.fillna(method="bfill").fillna(0)

/tmp/ipykernel_318/3525226833.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  sf_permits.fillna(method="bfill").fillna(0)


,Permit Number,Permit Type,Permit Type Definition,Permit Creation Date,Block,Lot,Street Number,Street Number Suffix,Street Name,Street Suffix,...,Existing Construction Type,Existing Construction Type Description,Proposed Construction Type,Proposed Construction Type Description,Site Permit,Supervisor District,Neighborhoods - Analysis Boundaries,Zipcode,Location,Record ID
0,201505065519,4,sign - erect,05/06/2015,0326,023,140,A,Ellis,St,...,3.0,constr type 3,1.0,constr type 1,Y,3.0,Tenderloin,94102.0,"(37.785719256680785, -122.40852313194863)",1380611233945
1,201604195146,4,sign - erect,04/19/2016,0306,007,440,A,Geary,St,...,3.0,constr type 3,1.0,constr type 1,Y,3.0,Tenderloin,94102.0,"(37.78733980600732, -122.41063199757738)",1420164406718
2,201605278609,3,additions alterations or repairs,05/27/2016,0595,203,1647,A,Pacific,Av,...,1.0,constr type 1,1.0,constr type 1,Y,3.0,Russian Hill,94109.0,"(37.7946573324287, -122.42232562979227)",1424856504716
3,201611072166,8,otc alterations permit,11/07/2016,0156,011,1230,A,Pacific,Av,...,5.0,wood frame (5),5.0,wood frame (5),Y,3.0,Nob Hill,94109.0,"(37.79595867909168, -122.41557405519474)",1443574295566
4,201611283529,6,demolitions,11/28/2016,0342,001,950,A,Market,St,...,3.0,constr type 3,1.0,constr type 1,Y,6.0,Tenderloin,94102.0,"(37.78315261897309, -122.40950883997789)",144548169992
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
198895,M862628,8,otc alterations permit,12/05/2017,0113,017A,1228,0,Montgomery,St,...,5.0,wood frame (5),5.0,wood frame (5),0,0.0,0,0.0,0,1489337276729
198896,201712055595,8,otc alterations permit,12/05/2017,0271,014,580,0,Bush,St,...,5.0,wood frame (5),5.0,wood frame (5),0,0.0,0,0.0,0,1489462354993
198897,M863507,8,otc alterations permit,12/06/2017,4318,019,1568,0,Indiana,St,...,0.0,0,0.0,0,0,0.0,0,0.0,0,1489539379952
198898,M863747,8,otc alterations permit,12/06/2017,0298,029,795,0,Sutter,St,...,0.0,0,0.0,0,0,0.0,0,0.0,0,1489608233656


Here, we used .fillna() twice inorder to fill the values with 0, that are still NaN even after the first back fill. This might happen if the last row is NaN.

Some notable .fillna() methods are
*   bfill - Fills the value of next row
*   ffill - Fills the value of previous row